src/LandIce/evaluators/LandIce_EnthalpyBasalResid_Def.hpp

```diff
@@ -50,25 +52,71 @@ EnthalpyBasalResid(const Teuchos::ParameterList& p, const Teuchos::RCP<Albany::L
   dl->node_vector->dimensions(dims);
   vecDimFO     = std::min((int)dims[2],2);
 
-  // Index of the nodes on the sides in the numeration of the cell
   Teuchos::RCP<shards::CellTopology> cellType;
   cellType = p.get<Teuchos::RCP <shards::CellTopology> > ("Cell Type");
   sideDim      = cellType->getDimension()-1;
-  sideNodes.resize(numSides);
-  for (int side=0; side<numSides; ++side)
-  {
-    // Need to get the subcell exact count, since different sides may have different number of nodes (e.g., Wedge)
+  int nodeMax = 0;
+  for (int side=0; side<numSides; ++side) {
     int thisSideNodes = cellType->getNodeCount(sideDim,side);
-    sideNodes[side].resize(thisSideNodes);
-    for (int node=0; node<thisSideNodes; ++node)
-    {
-      sideNodes[side][node] = cellType->getNodeMap(sideDim,side,node);
+    nodeMax = std::max(nodeMax, thisSideNodes);
+  }
+  sideNodes = Kokkos::DynRankView<int, PHX::Device>("sideNodes", numSides, nodeMax);
+  for (int side=0; side<numSides; ++side) {
+    int thisSideNodes = cellType->getNodeCount(sideDim,side);
+    for (int node=0; node<thisSideNodes; ++node) {
+      sideNodes(side,node) = cellType->getNodeMap(sideDim,side,node);
     }
   }
 
   this->setName("Enthalpy Basal Residual" + PHX::print<EvalT>());
 }
 
+template<typename EvalT, typename Traits, typename Type>
+KOKKOS_INLINE_FUNCTION
+void EnthalpyBasalResid<EvalT,Traits,Type>::
+operator() (const Enthalpy_Basal_Residual_Collapsed_Tag& tag, const int& sideSet_idx) const{
+  
+  const int cell = sideSet.elem_LID(sideSet_idx);
+  const int side = sideSet.side_local_id(sideSet_idx);
+
+  ScalarT val[4] = {0., 0., 0., 0.};
```

Comment:
Maybe instead of using directly 4, we could define a  variable  `maxNumNodesPerSide = 4`.

Reply:
We are already computing it at construction time `nodeMax`. All we need is to store it in the class. However, we then need to make sure we ask kokkos for some scratch memory for val, since we cannot allocate dynamically inside a kernel.

Reply:
Actually, we don't need val to be an array at all. We can fuse the two loops over numSideNodes, and make val into an automatic variable (so that Sacado does not evaluate it until it's time to store it in the resid).

Reply:
Let's change this to `constexpr int maxNumNodesPerSide = 4` for clarity. We can think about fusing loops later.